[TensorExpr] Nuke DepTracker and findAllNeededTensors.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55002 [TensorExpr] Add TORCH_APIs to all expr classes.
* #55001 [TensorExpr] Cleanup IRNodeType enum.
* #55000 [TensorExpr] Remove stale docs from DesignOverview.md.
* #54999 [TensorExpr] Nuke BaseCallNode.
* #54998 [TensorExpr] Nuke FunctionCall.
* **#54997 [TensorExpr] Nuke DepTracker and findAllNeededTensors.**

DepTracker was used to automatically pull in dependent computations from
output ones. While it seems quite convenient, it's led to several
architectural issues, which are fixed in this stack.

DepTracker worked on Tensors, which is a pair of Buf and Stmt. However,
Stmt could become stale and there was no way to reliably update the
corresponding tensor. We're now using Bufs and Stmts directly and moving
away from using Tensors to avoid these problems.

Removing DepTracker allowed to unify Loads and FunctionCalls, which
essentially were duplicates of each other.

Differential Revision: [D27446414](https://our.internmc.facebook.com/intern/diff/D27446414)

Commit:
[TensorExpr] Nuke DepTracker and findAllNeededTensors.

DepTracker was used to automatically pull in dependent computations from
output ones. While it seems quite convenient, it's led to several
architectural issues, which are fixed in this stack.

DepTracker worked on Tensors, which is a pair of Buf and Stmt. However,
Stmt could become stale and there was no way to reliably update the
corresponding tensor. We're now using Bufs and Stmts directly and moving
away from using Tensors to avoid these problems.

Removing DepTracker allowed to unify Loads and FunctionCalls, which
essentially were duplicates of each other.

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/tensorexpr/analysis.cpp:
```diff
@@ -1,72 +1,7 @@
-}
-
-} // namespace tensorexpr
+namespace tensorexpr {} // namespace tensorexpr
```
nit: Why not remove this file?

ZolotukhinM replied:
Good point!

navahgar reviewed (approved):
Glad to see that you got to it.

LGTM

Commit:
Update on "[TensorExpr] Nuke DepTracker and findAllNeededTensors."


DepTracker was used to automatically pull in dependent computations from
output ones. While it seems quite convenient, it's led to several
architectural issues, which are fixed in this stack.

DepTracker worked on Tensors, which is a pair of Buf and Stmt. However,
Stmt could become stale and there was no way to reliably update the
corresponding tensor. We're now using Bufs and Stmts directly and moving
away from using Tensors to avoid these problems.

Removing DepTracker allowed to unify Loads and FunctionCalls, which
essentially were duplicates of each other.

Differential Revision: [D27446414](https://our.internmc.facebook.com/intern/diff/D27446414)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Nuke DepTracker and findAllNeededTensors."


DepTracker was used to automatically pull in dependent computations from
output ones. While it seems quite convenient, it's led to several
architectural issues, which are fixed in this stack.

DepTracker worked on Tensors, which is a pair of Buf and Stmt. However,
Stmt could become stale and there was no way to reliably update the
corresponding tensor. We're now using Bufs and Stmts directly and moving
away from using Tensors to avoid these problems.

Removing DepTracker allowed to unify Loads and FunctionCalls, which
essentially were duplicates of each other.

Differential Revision: [D27446414](https://our.internmc.facebook.com/intern/diff/D27446414)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Nuke DepTracker and findAllNeededTensors."


DepTracker was used to automatically pull in dependent computations from
output ones. While it seems quite convenient, it's led to several
architectural issues, which are fixed in this stack.

DepTracker worked on Tensors, which is a pair of Buf and Stmt. However,
Stmt could become stale and there was no way to reliably update the
corresponding tensor. We're now using Bufs and Stmts directly and moving
away from using Tensors to avoid these problems.

Removing DepTracker allowed to unify Loads and FunctionCalls, which
essentially were duplicates of each other.

Differential Revision: [D27446414](https://our.internmc.facebook.com/intern/diff/D27446414)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Nuke DepTracker and findAllNeededTensors."


DepTracker was used to automatically pull in dependent computations from
output ones. While it seems quite convenient, it's led to several
architectural issues, which are fixed in this stack.

DepTracker worked on Tensors, which is a pair of Buf and Stmt. However,
Stmt could become stale and there was no way to reliably update the
corresponding tensor. We're now using Bufs and Stmts directly and moving
away from using Tensors to avoid these problems.

Removing DepTracker allowed to unify Loads and FunctionCalls, which
essentially were duplicates of each other.

Differential Revision: [D27446414](https://our.internmc.facebook.com/intern/diff/D27446414)

[ghstack-poisoned]
